File: Assets/Fireworks/Scripts/ParticleSystemSound.js
#pragma strict
var _shootSound:AudioClip[];

var _shootPitchMax:float = 1.25;
var _shootPitchMin:float = 0.75;
var _shootVolumeMax:float = 0.75;
var _shootVolumeMin:float = 0.25;

var _explosionSound:AudioClip[];
var _explosionPitchMax:float = 1.25;
var _explosionPitchMin:float = 0.75;
var _explosionVolumeMax:float = 0.75;
var _explosionVolumeMin:float = 0.25;

var _crackleSound:AudioClip[];
var _crackleDelay:float = .25;
var _crackleMultiplier:int = 3;
var _cracklePitchMax:float = 1.25;
var _cracklePitchMin:float = 0.75;
var _crackleVolumeMax:float = 0.75;
var _crackleVolumeMin:float = 0.25;

function LateUpdate() {
	var particles:ParticleSystem.Particle[]  = new ParticleSystem.Particle[particleSystem.particleCount];
	var length:int = particleSystem.GetParticles(particles);
	var i:int = 0;
	while (i < length){
 		if(_explosionSound.Length > 0 && particles[i].lifetime < Time.deltaTime){
 			SoundController.instance.Play(_explosionSound[Random.Range(0, _explosionSound.Length)], Random.Range(_explosionVolumeMax,_explosionVolumeMin), Random.Range(_explosionPitchMin,_explosionPitchMax), particles[i].position);
			if(_crackleSound.Length > 0){
			for(var j:int = 0; j <_crackleMultiplier; j++){
				Crackle(particles[i].position, _crackleDelay+j*.1);
			}
 		}
 		}
 		if(_shootSound.Length > 0 && particles[i].lifetime >= particles[i].startLifetime-Time.deltaTime){
 			SoundController.instance.Play(_shootSound[Random.Range(0, _shootSound.Length)], Random.Range(_shootVolumeMax,_shootVolumeMin), Random.Range(_shootPitchMin,_shootPitchMax), particles[i].position);
		}
		i++;
	}
}

function Crackle(pos:Vector3, delay:float){
	yield WaitForSeconds(delay);
	SoundController.instance.Play(_crackleSound[Random.Range(0, _crackleSound.Length)], Random.Range(_crackleVolumeMax,_crackleVolumeMin), Random.Range(_cracklePitchMax,_cracklePitchMin), pos);
}

File: Assets/Fireworks/Scripts/ParticleSystemSound.cs
using UnityEngine;
using System;
using System.Collections;


public class ParticleSystemSound:MonoBehaviour{
	public AudioClip[] _shootSound;
	
	public float _shootPitchMax = 1.25f;
	public float _shootPitchMin = 0.75f;
	public float _shootVolumeMax = 0.75f;
	public float _shootVolumeMin = 0.25f;
	
	public AudioClip[] _explosionSound;
	public float _explosionPitchMax = 1.25f;
	public float _explosionPitchMin = 0.75f;
	public float _explosionVolumeMax = 0.75f;
	public float _explosionVolumeMin = 0.25f;
	
	public AudioClip[] _crackleSound;
	public float _crackleDelay = .25f;
	public int _crackleMultiplier = 3;
	public float _cracklePitchMax = 1.25f;
	public float _cracklePitchMin = 0.75f;
	public float _crackleVolumeMax = 0.75f;
	public float _crackleVolumeMin = 0.25f;
	
	public void LateUpdate() {
		ParticleSystem.Particle[] particles  = new ParticleSystem.Particle[GetComponent<ParticleSystem>().particleCount];
		int length = GetComponent<ParticleSystem>().GetParticles(particles);
		int i = 0;
		while (i < length){
	 		if(_explosionSound.Length > 0 && particles[i].remainingLifetime < Time.deltaTime){
	 			SoundController.instance.Play(_explosionSound[UnityEngine.Random.Range(0, _explosionSound.Length)], UnityEngine.Random.Range(_explosionVolumeMax,_explosionVolumeMin), UnityEngine.Random.Range(_explosionPitchMin,_explosionPitchMax), particles[i].position);
				if(_crackleSound.Length > 0){
				for(int j = 0; j <_crackleMultiplier; j++){
					StartCoroutine(Crackle(particles[i].position, _crackleDelay+j*.1f));
				}
	 		}
	 		}
	 		if(_shootSound.Length > 0 && particles[i].remainingLifetime >= particles[i].startLifetime-Time.deltaTime){
	 			SoundController.instance.Play(_shootSound[UnityEngine.Random.Range(0, _shootSound.Length)], UnityEngine.Random.Range(_shootVolumeMax,_shootVolumeMin), UnityEngine.Random.Range(_shootPitchMin,_shootPitchMax), particles[i].position);
			}
			i++;
		}
	}
	
	public IEnumerator Crackle(Vector3 pos,float delay){
		yield return new WaitForSeconds(delay);
		SoundController.instance.Play(_crackleSound[UnityEngine.Random.Range(0, _crackleSound.Length)], UnityEngine.Random.Range(_crackleVolumeMax,_crackleVolumeMin), UnityEngine.Random.Range(_cracklePitchMax,_cracklePitchMin), pos);
	}
}
